lib/aws-chunked-decoder.js

```diff
@@ -0,0 +1,497 @@
+ *       +---------------------------+
+ *       | CRLF                      |  "\r\n"
+ *       +---------------------------+
+ *
+ * State Machine flow:
+ *
+ *                      (Start)
+ *                         |
+ *                         v
+ *                 +-------------------+
+ *                 |   Chunk-Size      |
+ *                 | [;chunk-signature]|
+ *                 +-------------------+
+ *                          |
+ *                          v
+ *                 +-------------------+
+ *                 |   CRLF (end line) |
+ *                 +-------------------+
+ *                          |
+ *                          v
+ *                 +-------------------+
+ *                 |   Chunk-Data      |  <-- payload bytes (N bytes)
+ *                 +-------------------+
+ *                          |
+ *                          v
+ *                 +-------------------+
+ *                 |   CRLF (end data) |
+ *                 +-------------------+
+ *                          |
+ *        +-----------------+----------------+
+ *        |                                  |
+ *        | (if size > 0)                    | (if size = 0)
+ *        v                                  v
+ *  (loop back to Chunk-Size)        +-------------------+
+ *                                   | TrailerHeaders    |
+ *                                   +-------------------+
+ *                                            |
+ *                                            v
+ *                                   +-------------------+
+ *                                   |   Final CRLF      |
+ *                                   +-------------------+
+ *                                            |
+ *                                            v
+ *                                         [END]
+ *  References:
+ *  - https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html
+ *  - RFC 9112 https://datatracker.ietf.org/doc/html/rfc9112
+ */
+
+/*
+ * AWS Chunked Encoding Decoder
+ *
+ * AWS S3 uses a special chunked encoding format when Content-Encoding is
+ * 'aws-chunked':
+ *
+ * Format:
+ * [chunk-size in hex];chunk-signature=[signature]\r\n
+ * [chunk-data]\r\n
+ * [chunk-size in hex];chunk-signature=[signature]\r\n
+ * [chunk-data]\r\n
+ * 0;chunk-signature=[final-signature]\r\n
+ * [optional trailing headers]\r\n
+ * \r\n
+ *
+ * This decoder strips the AWS chunk metadata and extracts only the actual data.
+ */
+function AwsChunkedDecoder(options) {
+    if (!(this instanceof AwsChunkedDecoder)) {
+        return (new AwsChunkedDecoder(options));
+    }
+    stream.Transform.call(this, options);
+    this._buffer = Buffer.alloc(0);
+    this._state = 'chunk-size';
+    // chunk-size, chunk-data, trailing-headers, complete
+    this._chunkSize = 0;
+    this._chunkBytesRead = 0;
+    this._totalBytesDecoded = 0;
+    this._log = options && options.log;
```

Comment:
Nit/Stylistic (feel free to disregard): We could `assert.object(options.log, 'options.log');`  here and skip the `if (this._log)...` conditionals all the places this class intends to log.